File: hphp/hack/src/hhbc/emit_method.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)

module SU = Hhbc_string_utils
module SN = Naming_special_names
module T = Aast
open Core_kernel
open Instruction_sequence

let from_ast_wrapper privatize make_name ast_class ast_method =
  let (pos, original_name) = ast_method.T.m_name in
  let class_name = SU.Xhp.mangle @@ Utils.strip_ns @@ snd ast_class.T.c_name in
  (* TODO: use something that can't be faked in user code *)
  let method_is_closure_body =
    original_name = "__invoke" && String.is_prefix ~prefix:"Closure$" class_name
  in
  let namespace = ast_class.T.c_namespace in
  let method_attributes =
    Emit_attribute.from_asts namespace ast_method.T.m_user_attributes
  in
  let method_attributes =
    if method_is_closure_body then
      method_attributes
    else
      Emit_attribute.add_reified_attribute
        method_attributes
        ast_method.T.m_tparams
  in
  let is_native = Hhas_attribute.has_native method_attributes in
  let is_native_opcode_impl =
    Hhas_attribute.is_native_opcode_impl method_attributes
  in
  let method_is_final = ast_method.T.m_final in
  let method_is_abstract =
    ast_class.T.c_kind = Ast_defs.Cinterface || ast_method.T.m_abstract
  in
  let method_is_static = ast_method.T.m_static in
  let method_visibility =
    if is_native_opcode_impl then
      Aast.Public
    else if privatize then
      Aast.Private
    else
      ast_method.T.m_visibility
  in
  let is_memoize =
    Emit_attribute.ast_any_is_memoize ast_method.T.m_user_attributes
  in
  let deprecation_info =
    if is_memoize then
      None
    else
      Hhas_attribute.deprecation_info method_attributes
  in
  let is_no_injection = Hhas_attribute.is_no_injection method_attributes in
  let ret = T.hint_of_type_hint ast_method.T.m_ret in
  let method_id = make_name ast_method.T.m_name in
  if not (method_is_static || method_is_closure_body) then
    List.iter ast_method.T.m_params ~f:(fun p ->
        if p.T.param_name = SN.SpecialIdents.this then
          Emit_fatal.raise_fatal_parse p.T.param_pos "Cannot re-assign $this");
  let method_is_async =
    ast_method.T.m_fun_kind = Ast_defs.FAsync
    || ast_method.T.m_fun_kind = Ast_defs.FAsyncGenerator
  in
  if
    ast_class.T.c_kind = Ast_defs.Cinterface
    && not (List.is_empty ast_method.T.m_body.T.fb_ast)
  then
    Emit_fatal.raise_fatal_parse
      pos
      (Printf.sprintf
         "Interface method %s::%s cannot contain body"
         class_name
         original_name);
  if
    (not method_is_static)
    && ast_class.T.c_final
    && ast_class.T.c_kind = Ast_defs.Cabstract
  then
    Emit_fatal.raise_fatal_parse
      pos
      ( "Class "
      ^ class_name
      ^ " contains non-static method "
      ^ original_name
      ^ " and therefore cannot be declared 'abstract final'" );
  let has_variadic_param =
    List.exists ast_method.T.m_params ~f:(fun p -> p.T.param_is_variadic)
  in
  if has_variadic_param && original_name = Naming_special_names.Members.__call
  then
    Emit_fatal.raise_fatal_parse pos
    @@ Printf.sprintf
         "Method %s::__call() cannot take a variadic argument"
         class_name;
  let default_dropthrough =
    if ast_method.T.m_abstract then
      Some
        (Emit_fatal.emit_fatal_runtimeomitframe
           pos
           ( "Cannot call abstract method "
           ^ class_name
           ^ "::"
           ^ original_name
           ^ "()" ))
    else
      None
  in
  let scope =
    [Ast_scope.ScopeItem.Method ast_method; Ast_scope.ScopeItem.Class ast_class]
  in
  let scope =
    if method_is_closure_body then
      Ast_scope.ScopeItem.Lambda (method_is_async, None) :: scope
    else
      scope
  in
  let closure_namespace =
    SMap.find_opt class_name (Emit_env.get_closure_namespaces ())
  in
  let namespace = Option.value closure_namespace ~default:namespace in
  let method_rx_level =
    match Rx.rx_level_from_ast ast_method.T.m_user_attributes with
    | Some l -> l
    | None ->
      if method_is_closure_body then
        Emit_env.get_lambda_rx_of_scope ast_class ast_method
      else
        Rx.NonRx
  in
  let (ast_body_block, is_rx_body, method_rx_disabled) =
    if method_rx_level <> Rx.NonRx then
      match Rx.halves_of_is_enabled_body ast_method.T.m_body with
      | Some (enabled_body, disabled_body) ->
        if Hhbc_options.rx_is_enabled !Hhbc_options.compiler_options then
          (enabled_body, true, false)
        else
          (disabled_body, false, true)
      | None -> (ast_method.T.m_body.T.fb_ast, true, false)
    else
      (ast_method.T.m_body.T.fb_ast, false, false)
  in
  let class_tparam_names =
    List.map ast_class.T.c_tparams.T.c_tparam_list (fun t -> snd t.T.tp_name)
  in
  let (method_body, method_is_generator, method_is_pair_generator) =
    if is_native_opcode_impl then
      ( Emit_native_opcode.emit_body
          scope
          namespace
          ast_class.T.c_user_attributes
          ast_method.T.m_name
          ast_method.T.m_params
          ret,
        false,
        false )
    else
      Emit_body.emit_body
        ~pos:ast_method.T.m_span
        ~scope
        ~is_closure_body:method_is_closure_body
        ~is_memoize
        ~is_native
        ~is_async:method_is_async
        ~is_rx_body
        ~debugger_modify_program:false
        ~deprecation_info
        ~skipawaitable:(ast_method.T.m_fun_kind = Ast_defs.FAsync)
        ~default_dropthrough
        ~return_value:instr_null
        ~namespace
        ~doc_comment:ast_method.T.m_doc_comment
        ~immediate_tparams:ast_method.T.m_tparams
        ~class_tparam_names
        ast_method.T.m_params
        ret
        [T.Stmt (Pos.none, T.Block ast_body_block)]
  in
  let method_is_interceptable =
    Interceptable.is_method_interceptable ast_class method_id
  in
  let method_span =
    if is_native_opcode_impl then
      (0, 0)
    else
      Hhas_pos.pos_to_span ast_method.T.m_span
  in
  let normal_function =
    Hhas_method.make
      method_attributes
      method_visibility
      method_is_static
      method_is_final
      method_is_abstract
      is_no_injection
      method_id
      method_body
      method_span
      method_is_async
      method_is_generator
      method_is_pair_generator
      method_is_closure_body
      method_is_interceptable
      is_memoize (*method_is_memoize_impl*)
      method_rx_level
      method_rx_disabled
  in
  [normal_function]

let from_ast class_ method_ =
  let is_memoize =
    Emit_attribute.ast_any_is_memoize method_.T.m_user_attributes
  in
  let make_name (_, name) =
    if is_memoize then
      Hhbc_id.Method.from_ast_name (name ^ Emit_memoize_helpers.memoize_suffix)
    else
      Hhbc_id.Method.from_ast_name name
  in
  from_ast_wrapper is_memoize make_name class_ method_

let from_asts class_ methods = List.concat_map methods ~f:(from_ast class_)

File: hphp/hack/src/hhbc/emit_method.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.
use ast_scope_rust::{self as ast_scope, Lambda, Scope, ScopeItem};
use emit_attribute_rust as emit_attribute;
use emit_body_rust as emit_body;
use emit_fatal_rust::{emit_fatal_runtimeomitframe, raise_fatal_parse};
use emit_memoize_helpers_rust as emit_memoize_helpers;
use emit_native_opcode_rust as emit_native_opcode;
use env::emitter::Emitter;
use hhas_attribute_rust as hhas_attribute;
use hhas_coeffects::HhasCoeffects;
use hhas_method_rust::{HhasMethod, HhasMethodFlags};
use hhas_pos_rust::Span;
use hhbc_id_rust::method;
use hhbc_string_utils_rust as string_utils;
use instruction_sequence::{instr, Result};
use naming_special_names_rust::{classes, special_idents, user_attributes};
use ocamlrep::rc::RcOc;
use options::{HhvmFlags, Options};
use oxidized::{ast as T, ast_defs};

use itertools::Either;
use std::borrow::Cow;

pub fn from_asts<'a>(
    emitter: &mut Emitter,
    class: &'a T::Class_,
    methods: &'a [T::Method_],
) -> Result<Vec<HhasMethod<'a>>> {
    methods
        .iter()
        .map(|m| from_ast(emitter, class, m))
        .collect::<Result<Vec<_>>>()
}

pub fn from_ast<'a>(
    emitter: &mut Emitter,
    class: &'a T::Class_,
    method_: impl Into<Cow<'a, T::Method_>>,
) -> Result<HhasMethod<'a>> {
    let method_: Cow<'a, T::Method_> = method_.into();
    let method = method_.as_ref();
    let is_memoize = method
        .user_attributes
        .iter()
        .any(|ua| user_attributes::is_memoized(&ua.name.1));
    let class_name = string_utils::mangle(string_utils::strip_global_ns(&class.name.1).into());
    let is_closure_body = &method.name.1 == "__invoke" && (class.name.1).starts_with("Closure$");
    let mut attributes =
        emit_attribute::from_asts(emitter, &class.namespace, &method.user_attributes)?;
    if !is_closure_body {
        attributes.extend(emit_attribute::add_reified_attribute(&method.tparams[..]));
    };
    let call_context = if is_closure_body {
        match &method.user_attributes[..] {
            [T::UserAttribute {
                name: ast_defs::Id(_, ref s),
                params: _,
            }] if s.eq_ignore_ascii_case("__DynamicMethCallerForce") => {
                Some("__SystemLib\\DynamicContextOverrideUnsafe".to_string())
            }
            _ => None,
        }
    } else {
        None
    };

    let is_native = attributes
        .iter()
        .any(|attr| attr.is(user_attributes::is_native));
    let is_native_opcode_impl = hhas_attribute::is_native_opcode_impl(&attributes);
    let is_abstract = class.kind.is_cinterface() || method.abstract_;
    let is_async = method.fun_kind.is_fasync() || method.fun_kind.is_fasync_generator();
    let is_no_injection = hhas_attribute::is_no_injection(&attributes);

    if !(method.static_ || is_closure_body) {
        for p in method.params.iter() {
            if p.name == special_idents::THIS {
                return Err(raise_fatal_parse(&p.pos, "Cannot re-assign $this"));
            }
        }
    };
    if class.kind.is_cinterface() && !method.body.ast.is_empty() {
        return Err(raise_fatal_parse(
            &method.name.0,
            format!(
                "Interface method {}::{} cannot contain body",
                class_name, &method.name.1
            ),
        ));
    };
    if !method.static_ && class.final_ && class.kind.is_cabstract() {
        return Err(raise_fatal_parse(
            &method.name.0,
            format!(
                "Class {} contains non-static method {} and therefore cannot be declared 'abstract final'",
                class_name, &method.name.1
            ),
        ));
    };

    let visibility = if is_native_opcode_impl {
        T::Visibility::Public
    } else if is_memoize {
        T::Visibility::Private
    } else {
        method.visibility
    };
    let deprecation_info = if is_memoize {
        None
    } else {
        hhas_attribute::deprecation_info(attributes.iter())
    };
    let default_dropthrough = if method.abstract_ {
        Some(emit_fatal_runtimeomitframe(
            &&method.name.0,
            format!(
                "Cannot call abstract method {}::{}()",
                class_name, &method.name.1
            ),
        ))
    } else {
        None
    };
    let mut scope = Scope {
        items: vec![
            ScopeItem::Class(ast_scope::Class::new_ref(class)),
            ScopeItem::Method(match &method_ {
                Cow::Borrowed(m) => ast_scope::Method::new_ref(&m),
                Cow::Owned(m) => ast_scope::Method::new_rc(&m),
            }),
        ],
    };
    if is_closure_body {
        scope.items.push(ScopeItem::Lambda(Lambda {
            is_async,
            coeffects: HhasCoeffects::default(),
        }))
    };
    let namespace = RcOc::clone(
        emitter
            .emit_global_state()
            .closure_namespaces
            .get(&class_name)
            .unwrap_or(&class.namespace),
    );
    let mut coeffects = HhasCoeffects::from_ast(&method.ctxs, &method.params);
    if method.ctxs == None && is_closure_body {
        let parent_coeffects = emitter
            .emit_global_state()
            .get_lambda_coeffects_of_scope(&class.name.1, &method.name.1);
        coeffects = parent_coeffects.inherit_to_child_closure()
    }
    if is_native_opcode_impl
        && (class.name.1.as_str() == classes::GENERATOR
            || class.name.1.as_str() == classes::ASYNC_GENERATOR)
    {
        match method.name.1.as_str() {
            "send" | "raise" | "throw" | "next" | "rewind" => {
                coeffects = coeffects.with_gen_coeffect()
            }
            _ => {}
        }
    }
    let (ast_body_block, is_rx_body, rx_disabled) = if !coeffects.is_any_rx() {
        (&method.body.ast, false, false)
    } else {
        match hhas_coeffects::halves_of_is_enabled_body(&method.body) {
            None => (&method.body.ast, true, false),
            Some((enabled_body, disabled_body)) => {
                if emitter
                    .options()
                    .hhvm
                    .flags
                    .contains(options::HhvmFlags::RX_IS_ENABLED)
                {
                    (enabled_body, true, false)
                } else {
                    (disabled_body, false, true)
                }
            }
        }
    };
    let (body, is_generator, is_pair_generator) = if is_native_opcode_impl {
        (
            emit_native_opcode::emit_body(
                emitter,
                &scope,
                namespace.as_ref(),
                &class.user_attributes,
                &method.name,
                &method.params,
                method.ret.1.as_ref(),
            )?,
            false,
            false,
        )
    } else {
        let class_tparam_names = class
            .tparams
            .iter()
            .map(|tp| (tp.name).1.as_str())
            .collect::<Vec<_>>();
        let mut flags = emit_body::Flags::empty();
        flags.set(
            emit_body::Flags::SKIP_AWAITABLE,
            method.fun_kind.is_fasync(),
        );
        flags.set(emit_body::Flags::MEMOIZE, is_memoize);
        flags.set(emit_body::Flags::CLOSURE_BODY, is_closure_body);
        flags.set(emit_body::Flags::NATIVE, is_native);
        flags.set(emit_body::Flags::RX_BODY, is_rx_body);
        flags.set(emit_body::Flags::ASYNC, is_async);
        flags.set(
            emit_body::Flags::HAS_COEFFECTS_LOCAL,
            coeffects.has_coeffects_local(),
        );
        emit_body::emit_body(
            emitter,
            namespace,
            Either::Right(ast_body_block),
            instr::null(),
            scope,
            emit_body::Args {
                immediate_tparams: &method.tparams,
                class_tparam_names: class_tparam_names.as_slice(),
                ast_params: &method.params,
                ret: method.ret.1.as_ref(),
                pos: &method.span,
                deprecation_info: &deprecation_info,
                doc_comment: method.doc_comment.clone(),
                default_dropthrough,
                call_context,
                flags,
            },
        )?
    };
    let name = {
        let mut name: method::Type = string_utils::strip_global_ns(&method.name.1)
            .to_string()
            .into();
        if is_memoize {
            name.add_suffix(emit_memoize_helpers::MEMOIZE_SUFFIX);
        };
        name
    };
    let is_interceptable = is_method_interceptable(emitter.options());
    let span = if is_native_opcode_impl {
        Span(0, 0)
    } else {
        Span::from_pos(&method.span)
    };
    let mut flags = HhasMethodFlags::empty();
    flags.set(HhasMethodFlags::IS_STATIC, method.static_);
    flags.set(HhasMethodFlags::IS_FINAL, method.final_);
    flags.set(HhasMethodFlags::IS_ABSTRACT, is_abstract);
    flags.set(HhasMethodFlags::IS_ASYNC, is_async);
    flags.set(HhasMethodFlags::IS_GENERATOR, is_generator);
    flags.set(HhasMethodFlags::IS_PAIR_GENERATOR, is_pair_generator);
    flags.set(HhasMethodFlags::IS_CLOSURE_BODY, is_closure_body);
    flags.set(HhasMethodFlags::IS_INTERCEPTABLE, is_interceptable);
    flags.set(HhasMethodFlags::IS_MEMOIZE_IMPL, is_memoize);
    flags.set(HhasMethodFlags::RX_DISABLED, rx_disabled);
    flags.set(HhasMethodFlags::NO_INJECTION, is_no_injection);
    Ok(HhasMethod {
        attributes,
        visibility,
        name,
        body,
        span,
        coeffects,
        flags,
    })
}

fn is_method_interceptable(opts: &Options) -> bool {
    opts.hhvm
        .flags
        .contains(HhvmFlags::JIT_ENABLE_RENAME_FUNCTION)
}
